File: docs/validation/python/app/scenarios/av08_version_fidelity.py
import tempfile
from pathlib import Path
from machine_config import MachineConfigReader, MachineConfigWriter

"""
AV-08: Roundtrip version string fidelity
ID:          AV-08
Title:       File_Version string survives write→read unchanged
Category:    adapter-versioning
Layer:       writer + reader
Precondition: fixtures/reference_config.h5
Action:      Read → write to temp → read temp → compare File_Version strings
Expected:    File_Version is exactly "1.0" after roundtrip in all five languages.
             No whitespace added, no truncation, no case change.
Rationale:   A writer that silently normalizes the version string would break
             future version dispatch for files it produces.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    path = Path(fixtures_dir) / "reference_config.h5"
    cfg = MachineConfigReader(path).parse()
    orig_version = cfg.meta.file_version.strip()

    with tempfile.NamedTemporaryFile(suffix=".h5", delete=False) as f:
        tmp = f.name

    MachineConfigWriter(cfg).write(tmp)
    rb = MachineConfigReader(tmp).parse()
    rb_version = rb.meta.file_version.strip()

    if rb_version != "1.0":
        return False, f"file_version after roundtrip: expected '1.0', got '{rb_version}'"
    if rb_version != orig_version:
        return False, f"file_version changed: '{orig_version}' → '{rb_version}'"

    return True, f"File_Version survives roundtrip unchanged: '{rb_version}'"

File: docs/validation/cpp/app/src/scenarios/av08_version_fidelity.cpp
// AV-08: File_Version string survives write -> read unchanged
#include "av08_version_fidelity.hpp"
#include "machine_config/machine_config.hpp"

namespace av08 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path& fixturesDir, const std::filesystem::path&) {
    auto path = fixturesDir / "reference_config.h5";
    try {
        MachineConfig cfg = MachineConfigReader{path}.parse();
        std::string origVersion = cfg.meta.file_version;

        auto tmp = scenarios::makeTempPath(".h5");
        MachineConfigWriter{cfg}.write(tmp);
        MachineConfig rb = MachineConfigReader{tmp}.parse();
        std::string rbVersion = rb.meta.file_version;

        if (rbVersion != "1.0") {
            return {false, "file_version after roundtrip: expected '1.0', got '" + rbVersion + "'"};
        }
        if (rbVersion != origVersion) {
            return {false, "file_version changed: '" + origVersion + "' -> '" + rbVersion + "'"};
        }

        return {true, "File_Version survives roundtrip unchanged: '" + rbVersion + "'"};
    } catch (const std::exception& e) {
        return {false, std::string("exception: ") + e.what()};
    }
}

} // namespace av08
